modules/ding_campaign_plus/modules/ding_campaign_plus_basic/ding_campaign_plus_basic.module

```diff
@@ -0,0 +1,543 @@
+<?php
+/**
+ * @file
+ * Defines basic campaigns triggers and handling.
+ */
+
+define('DING_CAMPAIGN_PLUS_BASIC_TYPE', 'basic');
+
+/**
+ * Implements hook_ding_campaign_plus_info().
+ */
+function ding_campaign_plus_basic_ding_campaign_plus_info() {
+  return array(
+    'title' => t('Basic'),
+    'type' => DING_CAMPAIGN_PLUS_BASIC_TYPE,
+    'form' => 'ding_campaign_plus_basic_admin_form',
+    'auto' => 'ding_campaign_plus_basic_auto_admin_form',
+    'weight' => 5,
+  );
+}
+
+/**
+ * Implements hook_ding_campaign_plus_default_weights().
+ */
+function ding_campaign_plus_basic_ding_campaign_plus_default_weights() {
+  $types = _ding_campaign_plus_basic_get_types();
+
+  $weight = -10;
+  $weights = array();
+  foreach ($types as $key => $type) {
+    $weights[$key] = array(
+      'prefix' => t('Basic'),
+      'title' => $type,
+      'weight' => $weight++,
+    );
+  }
+
+  return $weights;
+}
+
+/**
+ * Implements hook_menu().
+ */
+function ding_campaign_plus_basic_menu() {
+  $items = array();
+
+  $items['ding_campaign_plus/basic/autocomplete/%'] = array(
+    'title' => 'Campaign rule autocomplete',
+    'description' => 'Auto complete field for campaign rule values.',
+    'page callback' => '_ding_campaign_plus_basic_admin_autocomplete',
+    'page arguments' => array(3),
+    'access arguments' => array('access campaign plus triggers'),
+    'type' => MENU_CALLBACK,
+  );
+
+  return $items;
+}
+
+/**
+ * Implements hook_node_load().
+ */
+function ding_campaign_plus_basic_node_load($nodes, $types) {
+  foreach ($nodes as $node) {
+    if ($node->type === 'ding_campaign_plus') {
+      $rules = db_select('ding_campaign_plus_basic', 'dcpb')
+        ->fields('dcpb')
+        ->condition('nid', $node->nid)
+        ->execute()
+        ->fetchAllAssoc('id');
+
+      if (!isset($node->ding_campaign_plus)) {
+        $node->ding_campaign_plus = array();
+      }
+      $node->ding_campaign_plus += $rules;
+    }
+  }
+}
+
+/**
+ * Implements hook_node_insert().
+ *
+ * Save campaign basic rules for a given campaign.
+ */
+function ding_campaign_plus_basic_node_insert($node) {
+  if ($node->type === 'ding_campaign_plus') {
+    $triggers = $node->tabs[DING_CAMPAIGN_PLUS_BASIC_TYPE];
+
+    foreach ($triggers['rules'] as $rule) {
+      if (!empty($rule['value'])) {
+        $value = $rule['value'];
+        switch ($rule['type']) {
+          case 'path':
+            $value = $rule['value'];
+            break;
+
+          default:
+            // Parse the value to get faster look-up when looking up campaigns
+            // to display.
+            preg_match('/.+\[(\d+)\]$/', $value, $matches);
+            $value = is_array($matches) ? end($matches) : '';
```

Comment:
`$matches` will be empty if `$value` does not match the regular expression. Please rewrite to make it clear that we know/handle this correctly, e.g.

```
$value = preg_match('/.+\[(?<value>\d+)\]$/', $value, $matches) ? $matches['value'] : '';
```